lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3InputStream.java

```diff
@@ -0,0 +1,268 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.filesystem.s3;
+
+import io.trino.filesystem.Location;
+import io.trino.filesystem.TrinoInputStream;
+import software.amazon.awssdk.core.ResponseInputStream;
+import software.amazon.awssdk.core.exception.AbortedException;
+import software.amazon.awssdk.core.exception.SdkException;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.GetObjectRequest;
+import software.amazon.awssdk.services.s3.model.GetObjectResponse;
+import software.amazon.awssdk.services.s3.model.NoSuchKeyException;
+
+import java.io.EOFException;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InterruptedIOException;
+
+import static java.lang.Math.max;
+import static java.util.Objects.requireNonNull;
+
+final class S3InputStream
+        extends TrinoInputStream
+{
+    private static final int MAX_SKIP_BYTES = 1024 * 1024;
+
+    private final Location location;
+    private final S3Client client;
+    private final GetObjectRequest request;
+    private final Long length;
+
+    private boolean closed;
+    private ResponseInputStream<GetObjectResponse> in;
+    private long streamPosition;
+    private long nextReadPosition;
+
+    public S3InputStream(Location location, S3Client client, GetObjectRequest request, Long length)
+    {
+        this.location = requireNonNull(location, "location is null");
+        this.client = requireNonNull(client, "client is null");
+        this.request = requireNonNull(request, "request is null");
+        this.length = length;
+    }
+
+    @Override
+    public int available()
+            throws IOException
+    {
+        ensureOpen();
+        return 0;
```

Comment:
Not sure this is needed, but can you return `in.available()` here?

Reply:
I see `getAvailable` below, you should explain why that isn't being used here

Reply:
It won't work if there was a seek and we haven't repositioned the underlying stream, but I can handle that